cpp/src/datetime/datetime_ops.cu

```diff
@@ -68,98 +67,152 @@ struct extract_component_operator {
   }
 };
 
-template <datetime_component Component>
-struct launch_extract_component {
+// Round up the date to the last day of the mont and return the
+// date only (without the time component)
+struct extract_last_day {
+  template <typename Timestamp>
+  CUDA_DEVICE_CALLABLE timestamp_D operator()(Timestamp const ts) const {
+    using namespace simt::std::chrono;
+    // IDEAL: does not work with CUDA10.0 due to nvcc compiler bug
+    // cannot invoke ym_last_day.day()
+    //const year_month_day orig_ymd(floor<days>(ts));
+    //const year_month_day_last ym_last_day(orig_ymd.year(), month_day_last(orig_ymd.month()));
+    //return timestamp_D(sys_days(ym_last_day));
+
+    // Only has the days - time component is chopped off, which is what we want
+    auto days_since_epoch = floor<days>(ts);
+
+    const auto chrono_year = year_month_day(days_since_epoch).year();
+
+    const auto chrono_month = year_month_day(days_since_epoch).month();
+    const unsigned month_val = static_cast<unsigned>(chrono_month);
+
+    static const uint8_t days_in_month[] = {31, 28, 31, 30, 31, 30,
+                                            31, 31, 30, 31, 30, 31};
+    day chrono_day(days_in_month[month_val - 1]);
+    // Handle leap years
+    if (chrono_month == month(February) && chrono_year.is_leap()) {
+      chrono_day++;
+    }
+    unsigned orig_day(year_month_day(days_since_epoch).day());
+    unsigned new_day(chrono_day);
+    days_since_epoch += days(new_day - orig_day);
+
+    return timestamp_D(days_since_epoch);
```

Comment:
There are a lot of unnecessary temporary variables being used here which hurt readability. We can reduce this to 4 lines of code:
```cpp
    // Only has the days - time component is chopped off, which is what we want
    auto const days_since_epoch = floor<days>(ts);
    auto const date             = year_month_day(days_since_epoch);
    auto const last_day         = detail::last_date(date.month(), date.year().is_leap());

    return timestamp_D(days_since_epoch + last_day - date.day());
```
If we write a more declarative helper function something like this:
```cpp
auto last_day(std::chrono::month month, bool is_leap_year) -> uint8_t {
    switch (month) {
        case January   : return 31;
        case February  : return is_leap_year ? 29 : 28;  
        case March     : return 31;
        case April     : return 30;
        case May       : return 31;  
        case June      : return 30;
        case July      : return 31;
        case August    : return 31;
        case September : return 30; 
        case October   : return 31;
        case November  : return 30;
        case December  : return 31; 
        default: // handle error;
    }
}
```

Reply:
Can't this all just be a one-liner using `std::chrono::year_month_day_last`? 

https://en.cppreference.com/w/cpp/chrono/year_month_day_last

Reply:
@codereport - i never wanted to pursue the long version and i'd the original 3 liner commented just before this. i'd to try various concise approaches to circumvent the compiler issue, and i gave up eventually before quickly coming up with a version.

hopefully, this will be thrown out eventually in favor of the much better alternative `simt::std::chrono::year_month_day_last`